services/src/main/java/org/keycloak/social/apple/AppleIdentityProvider.java

```diff
@@ -0,0 +1,261 @@
+            JsonNode encodedIdToken = mapper.readTree(response).get("id_token");
+
+            JsonWebToken idToken = validateToken(encodedIdToken.asText(), false);
+
+            BrokeredIdentityContext user = new BrokeredIdentityContext(idToken.getSubject());
+            user.setUsername(idToken.getSubject());
+            if (userData !=null) {
+                JsonNode profile = mapper.readTree(userData);
+                String email = profile.get("email").asText();
+
+                String firstName = profile.get("name").get("firstName").asText();
+                String lastName = profile.get("name").get("lastName").asText();
+
+                user.setEmail(email);
+                user.setFirstName(firstName);
+                user.setLastName(lastName);
+                user.setUsername(firstName + " " + lastName);
+                AbstractJsonUserAttributeMapper.storeUserProfileForMapper(user, profile, getConfig().getAlias());
+            }
+
+            return user;
+        }
+
+        public SimpleHttp generateTokenRequest(String authorizationCode) {
+            KeycloakContext context = session.getContext();
+            SimpleHttp tokenRequest = SimpleHttp.doPost(getConfig().getTokenUrl(), session)
+                    .param(OAUTH2_PARAMETER_CODE, authorizationCode)
+                    .param(OAUTH2_PARAMETER_REDIRECT_URI, Urls.identityProviderAuthnResponse(context.getUri().getBaseUri(),
+                            getConfig().getAlias(), context.getRealm().getName()).toString())
+                    .param(OAUTH2_PARAMETER_GRANT_TYPE, OAUTH2_GRANT_TYPE_AUTHORIZATION_CODE);
+            return authenticateTokenRequest(tokenRequest);
+        }
+    }
+
+    private String generateJWS(String p8Content, String keyId, String teamId) {
+        try {
+            KeyFactory kf = KeyFactory.getInstance("ECDSA");
+            PKCS8EncodedKeySpec keySpecPKCS8 = new PKCS8EncodedKeySpec(
+                    Base64.getDecoder().decode(
+                            p8Content
+                                    .replaceAll("-----BEGIN PRIVATE KEY-----", "")
+                                    .replaceAll("-----END PRIVATE KEY-----", "")
+                                    .replaceAll("\\n", "")
+                                    .replaceAll(" ", "")
+                    ));
+            PrivateKey privateKey = kf.generatePrivate(keySpecPKCS8);
+            KeyWrapper keyWrapper = new KeyWrapper();
+            keyWrapper.setAlgorithm("ES256");
+            keyWrapper.setPrivateKey(privateKey);
+
+            return new JWSBuilder()
+                    .kid(keyId)
+                    .jsonContent(generateClientToken(teamId))
+                    .sign(new ServerECDSASignatureSignerContext(keyWrapper));
+        } catch (Exception e) {
+            logger.error("Unable to generate JWS");
+        }
+        return null;
+    }
+
+    private boolean isValidSecret(String clientSecret) {
```

Comment:
Don't you need to check whether or not `clientSecret` is null ? It should avoid the `JWSInputException` in that case and only catch errors when the JWS is malformed.

I would also change the logger to DEBUG.